File: src/org/genyris/classification/ClassMROComparator.java
package org.genyris.classification;

import java.util.Comparator;

import org.genyris.core.Dictionary;
import org.genyris.core.Exp;
import org.genyris.core.SimpleSymbol;
import org.genyris.exception.AccessException;
import org.genyris.interp.UnboundException;

public class ClassMROComparator implements Comparator {

	private SimpleSymbol NIL, SUPERCLASSES;
    public ClassMROComparator(SimpleSymbol NIL, SimpleSymbol superclasses) {
       this.NIL = NIL;
       this.SUPERCLASSES = superclasses;
    }

    public int compare(Object o1, Object o2) {
        Dictionary c1 = (Dictionary) o1;
        Dictionary c2 = (Dictionary) o2;
        return getClassDepth(c1) - getClassDepth(c2);
    }
    private int getClassDepth(Exp klass) {
        Dictionary c1 = (Dictionary) klass;
        try {
            Exp superclasses = c1.lookupVariableShallow(SUPERCLASSES);
            if(superclasses == NIL) {
                return 0;
            }
            int retval = 0;
            while( superclasses != NIL) {
                int tmp = 1 + getClassDepth(superclasses.car());
                if (retval < tmp ){
                    retval = tmp;
                }
                superclasses = superclasses.cdr();
            }
            return retval;
        } catch (UnboundException e) {
            return 0;
        } catch (AccessException e) {
            return 1000000;
        }
    }
}

File: src/org/genyris/classification/ClassMROComparator.kt
package org.genyris.classification

import org.genyris.core.Dictionary
import org.genyris.core.Exp
import org.genyris.core.SimpleSymbol
import org.genyris.exception.AccessException
import org.genyris.interp.UnboundException
import kotlin.Any
import kotlin.Comparator
import kotlin.Int

class ClassMROComparator(private val NIL: SimpleSymbol?, private val SUPERCLASSES: SimpleSymbol?) : Comparator<Any?> {
    override fun compare(o1: Any?, o2: Any?): Int {
        val c1 = o1 as Dictionary?
        val c2 = o2 as Dictionary?
        return getClassDepth(c1) - getClassDepth(c2)
    }

    private fun getClassDepth(klass: Exp?): Int {
        val c1 = klass as Dictionary
        try {
            var superclasses = c1.lookupVariableShallow(SUPERCLASSES)
            if (superclasses === NIL) {
                return 0
            }
            var retval = 0
            while (superclasses !== NIL) {
                val tmp = 1 + getClassDepth(superclasses.car())
                if (retval < tmp) {
                    retval = tmp
                }
                superclasses = superclasses.cdr()
            }
            return retval
        } catch (e: UnboundException) {
            return 0
        } catch (e: AccessException) {
            return 1000000
        }
    }
}
